bitcoin/test/fuzz/miniscript_decode.cpp

```diff
@@ -0,0 +1,163 @@
+        std::copy(first, last, keyid.begin());
+        const auto it = dummy_keys_map.find(keyid);
+        if (it == dummy_keys_map.end()) return false;
+        key = it->second;
+        return true;
+    }
+
+    // Timelock challenges satisfaction. Make the value (deterministically) vary to explore different
+    // paths.
+    bool CheckAfter(uint32_t value) const { return value % 2; }
+    bool CheckOlder(uint32_t value) const { return value % 2; }
+
+    // Signature challenges fulfilled with a dummy signate, if it was one of our dummy keys.
+    miniscript::Availability Sign(const CPubKey& key, std::vector<unsigned char>& sig) const {
+        const auto it = dummy_sigs.find(key);
+        if (it == dummy_sigs.end()) return miniscript::Availability::NO;
+        sig = it->second;
+        return miniscript::Availability::YES;
+    }
+
+    //! Lookup generalization for all the hash satisfactions below
+    miniscript::Availability LookupHash(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage,
+                                        const std::map<std::vector<unsigned char>, std::vector<unsigned char>>& map) const
+    {
+        const auto it = map.find(hash);
+        if (it == map.end()) return miniscript::Availability::NO;
+        preimage = it->second;
+        return miniscript::Availability::YES;
+    }
+    miniscript::Availability SatSHA256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return LookupHash(hash, preimage, sha256_preimages);
+    }
+    miniscript::Availability SatRIPEMD160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return LookupHash(hash, preimage, ripemd160_preimages);
+    }
+    miniscript::Availability SatHASH256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return LookupHash(hash, preimage, hash256_preimages);
+    }
+    miniscript::Availability SatHASH160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return LookupHash(hash, preimage, hash160_preimages);
+    }
+};
+
+Satisfier SATISFIER;
+
+void initialize_miniscript_decode() {
+    ECC_Start();
+    SATISFIER.Init();
+}
+
+FUZZ_TARGET_INIT(miniscript_decode, initialize_miniscript_decode)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    const std::optional<CScript> script = ConsumeDeserializable<CScript>(fuzzed_data_provider);
+    if (!script) return;
+
+    const auto ms = miniscript::FromScript(*script, SATISFIER);
+    if (ms) {
+        // We can roundtrip it to its string representation.
+        std::string ms_str;
+        assert(ms->ToString(SATISFIER, ms_str));
+        assert(*miniscript::FromScript(*script, SATISFIER) == *ms);
+        // The Script representation must roundtrip.
+        const CScript ms_script = ms->ToScript(SATISFIER);
+        assert(ms_script == *script);
```

Comment:
That's wrong: i had noticed it's not actually the case when coding the Python implem, but forgot about it. It must however be pretty rare as i ran this target a decent amount of time, found a number of other bugs but this assert never failed.

Reply:
Aaand in the end i actually think it is *right*. It's because this target works by first decoding from Script. I think:
- This doesn't hold: "encoding a Ms to Script and then decoding it must give the same representation"
- But this does: "decoding a Script to Ms and the re-encoding it must give the same script"